### ISSUE TITLE
Create keyboard shortcut to focus search bar

### ISSUE THREAD

**Issue by nlhkabu** [pypa/warehouse#6099]
**What's the problem this feature will solve?**
We have decided *not* to autofocus the search bar on the PyPI homepage (see #6088)

However, we should consider creating a keyboard shortcut to move users to the search bar.

**Describe the solution you'd like**
The solution should be:

- Easy for PyPI users to discover
- Simple
- It should *not* interfere with keyboard shortcuts used by screenreader users (we need to research this)

**Additional context**
Suggestion from @amcgregor:

On YouTube typing `/` activates search, while typing `?` shows a shortcut reference.  See the [original comment](https://github.com/pypa/warehouse/issues/3440#issuecomment-505631589) for more info.

**Comment by amcgregor** [comment]
Here's a short story summary for this, if this helps in any way.  :)

---

As a Python developer attempting to find back ("jump to") a specific project, or looking for a package that meets my requirements, I want to be able to rapidly search so that I can get on with the real work.

### Discoverability

**Given** a _new visitor_ to the package index,
**And** they haven not seen the hint yet (i.e. this is their first visit),
**Then** a _hint should be temporarily shown_ suggesting `/` (for quick search) and `?` to discover all available shortcuts.

**Given** a _visitor to the package index unaware of the availability of shortcuts_,
**When** the user _presses keys not assigned to shortcuts_,
**Then** a _hint should be temporarily shown_ suggesting `/` (for quick search) and `?` to discover all available shortcuts.

### Proposed Involved Shortcuts

**Given** a _visitor_ to the package index,
**When** that visitor _presses_ `/`,
**Then** the _search field_ should become focused.

**Given** a _visitor_ to the package index,
**When** that user _presses_ `?`,
**Then** a _popover reference of shortcuts_ should become visible.

---

Rationale: a growing de-facto standard across multiple sites, e.g. GitHub, YouTube, where `/` replicates the command-mode Vim shortcut for search.  Reddit has keyboard shortcuts for most navigation and actions (voting, replying, saving, hiding, collapsing/expanding), with `Q` as the shortcut for accessing search-like filters, preserving `?` as the discoverability option.  Additional discovery of sites involving shortcuts like these is somewhat hampered by the generic nature of the terms, "websites using keyboard shortcuts" finds nothing but keyboard shortcut references for Windows and things like that.  ;)

**Comment by nlhkabu** [comment]
Note: WCAG 2.1 - Criteria 2.1.4:

>  If a keyboard shortcut uses printable character keys, then the user must be able to
disable the key command, change the defined key to a non-printable key (Ctrl, Alt, etc.),
or only activate the shortcut when an associated interface component or button is
focused.

**Comment by amcgregor** [comment]
Interesting.  GitHub and YouTube fail this purely on the grounds of non-configurability, if interpreted _very_ strictly.  Alternatively, because the search navigation and discovery keys are actually outside the word-character range, and if targeted such that pressing those while focused on an input of some kind won't trigger—and typing into the browser's page search won't, it will require focus on or within the document—then I would argue "only activate the shortcut when an associated interface component or button is focused" is met.  Jump-to-search (and show-help) are global, page-wide actions, only requiring focus within the document.  (Unlike, say, pressing enter, which is contextual based on specific focus, i.e. activating a button or link contained within the page.)

Allowing for configuration or customization of shortcuts can be great for power users, however it is also a great way to vastly increase the support overhead for the feature, as it will no longer provide a uniform or consistent interface, documentation may become inconsistent with a user's reality, … though a global account-wide toggle for it would be entirely sensible!  (Though, not logged in, no shortcuts, period?)

**Comment by amcgregor** [comment]
@saippuakauppias "Not logical", but consistent for anyone prone to using keyboard shortcuts extensively outside the Microsoft world.  Open a terminal, `less /etc/php.ini.default`, type a forward slash… and now you're searching.  `vim` a file, type `/`, and now you're searching.  <kbd>⌃F</kbd> / <kbd>⌃G</kbd> is a ultimately Microsoft-ism, and even that's inconsistent: <kbd>F3</kbd> is nearly universal, <kbd>⌥D</kbd> is not uncommon for focusing an omnibar/address bar…  (`nano`? <kbd>⌃W</kbd>, possibly for "where") — though most sites I come across that are based primarily on a search workflow for primary use simply auto-focus the bar, straight.  (Another example to add to the pile of those: Craigslist.)

Ultimately I, personally, use Pypi in one of two ways: either I already know the exact name of the package I'm interested in and let my browser auto-complete an existing package URI, ⌥⌫ (delete prior word) that name, type a new one, and directly navigate to my destination, or I'm looking for an appropriate package given tags or a general sense of a solution to look for, and I navigate to the site root with the sole intent of searching.  Occasionally I might even release a package update and need to verify it… but I do that by directly navigating to the package released.

Pressing tab 7 times, exactly, or either hesitate or fumble with toggling shift to go back… or arbitrarily remember I can press tab once, then enter to "jump to content", then tab one more time… these "currently available solutions for jumping to the search" only keep getting worse as I dig.  All for the sake of not setting `autofocus` or avoiding a commonly understood and widely deployed shortcut.  Literally type <kbd>/</kbd> or <kbd>?</kbd> whilst looking at this ticket.

_What is lost by setting `autofocus`?_  I'm seeing some WCAG discussion on that last referenced issue, but the phrasing seems off.  Additionally, from a user perspective, 8.5% seems like splitting hairs as to not being "nearly all", given those self-same statistics indicate it would clearly benefit 52.9% of users, a majority—and given the remaining 38.6% **just leave**.

**What exact breakdown of "other" involves an action whose navigation would be bypassed by that autofocus?**  And that given a clear indication of focus (which is already a WCAG requirement for accessible field navigation… there is a border inlay, if not truly obviously clear) providing a grounding for direction of navigation to the desired element, e.g. visually indicating the need for reverse traversal upwards from that element.

The bureaucracy behind such a "simple" feature request is as intriguing as it is frustrating—and I mean that with all good humour.  I'm exceedingly patient on things like this.  😉  Thank you.

**Comment by amcgregor** [comment]
I very much appreciate the addition of this functionality, thank you.

In the intervening time since initially requesting, since I saw repeated rejections of functionality to improve the usability situation… I adapted my browser itself to assist me with this. `py` <kbd>⇥</kbd> <samp>[package name]</samp> <kbd>⏎</kbd> automatically submits whatever is after `py` appended to `https://pypi.org/project/` (it thinks that's the "base search path" for a provider named `py`.) Don't even need to hit a landing page first. Or any page. 😉
